timemachine/fe/energy_decomposition.py

```diff
@@ -93,11 +98,13 @@ def compute_energy_decomposed_u_kln(states: List[EnergyDecomposedState]) -> np.n
         assert len(state.batch_u_fns) == n_components
 
     u_kln_by_component = np.zeros((n_components, K, K, n_frames))
-    for comp in range(n_components):
+    for l in range(K):
+        # Load the frames into memory, then evaluate all of the components
+        # Done to avoid repeatedly reading from disk
+        xs, boxes = np.array([x for x in states[l].frames]), states[l].boxes
```

Comment:
Seems like this could get pretty large in memory for solvent and complex legs?

Reply:
We are already loading this into memory, by doing this we load it once rather than every call to every potential

Reply:
We are still incurring multiple copies of the coordinates to the GPU though. A more performance fix would be to allow running a batch of potentials on coordinates and getting out the per potential energies

Reply:
Sounds like another argument for wrapped `DeviceBuffer` :stuck_out_tongue:

Reply:
> We are already loading this into memory

Ah, right, didn't read carefully enough